File: String/68_TextJustification.py
class Solution(object):
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        res = []
        nums = []
        n = len(words)
        for word in words:
            nums.append(len(word))
        newStart = 0
        currLen = nums[0]
        for i in range(1, n):
            if currLen + 1 + nums[i] <= maxWidth:
                currLen = currLen + 1 + nums[i]
            else:
                # bigger than maxWidth
                res.append(self.buildString(newStart, i - 1, maxWidth, currLen, words))
                newStart = i
                currLen = nums[i]
        res.append(self.buildStringLastline(newStart, n - 1, maxWidth, currLen, words))
        return res
        
    def buildStringLastline(self, start, end, maxWidth, currLen, words):
        gap = end - start
        spaceNum = maxWidth - currLen
        s = []
        for i in range(start, end):
            s.append(words[i] + " ")
        s.append(words[end] + self.buildSpaces(spaceNum))
        return "".join(s)
    
    def buildString(self, start, end, maxWidth, currLen, words):
        gap = end - start
        if gap == 0:
            return words[start] + self.buildSpaces(maxWidth - currLen)
        spaceNum = maxWidth - currLen + gap
        ave = spaceNum / gap
        remain = spaceNum % gap
        avespaces = self.buildSpaces(ave)
        s = []
        for i in range(start, end):
            s.append(words[i])
            if remain > 0:
                s.append(avespaces + " ")
                remain -= 1
            else:
                s.append(avespaces)
        s.append(words[end])
        return "".join(s)

    def buildSpaces(self, n):
        s = []
        for i in range(n):
            s.append(" ")
        return "".join(s)

File: 68. Text Justification.java
public class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> res = new ArrayList<>();
        System.out.println(words.length);
        if (words == null || words.length == 0) return res;
        ArrayList<String> list = new ArrayList<String>();
        int len = 0;
        int i = 0;
        while (i < words.length) {
            len += list.size() == 0 ? words[i].length() : words[i].length() + 1;
            // System.out.println(len);
            if (len == maxWidth) {
                list.add(words[i]);
                res.add(String.join(" ", list));
                i++;
                len = 0;
                list.clear();
            } else if (len > maxWidth) {
                justfy(res, list, maxWidth);
                len = 0;
                list.clear();
            } else {
                list.add(words[i]);
                if (i == words.length - 1) {
                    StringBuilder sb = new StringBuilder();
                    int gap = maxWidth - len;
                    for (int j = 0; j < gap; j++) {
                        sb.append(" ");
                    }
                    res.add(String.join(" ", list) + sb.toString());
                }
                i++;
            }
        }
        return res;
    }

    public void justfy(List<String> res, List<String> list, int maxWidth) {
        // This    is   ban
        if (list.size() == 1) {
            StringBuilder sb = new StringBuilder();
            sb.append(list.get(0));
            int gap = maxWidth - sb.length();
            for (int i = 0; i < gap; i++) {
                sb.append(" ");
            }
            res.add(sb.toString());
            return;
        } 

        int len = 0;
        for (String s : list) {
            len += s.length();
        }
        int remainlen = maxWidth - len;
        int m = list.size() - 1;
        int cnt = remainlen / m;
        int remains = remainlen % m;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < m; i++) {
            sb.append(list.get(i));
            for (int j = 0; j < cnt; j++) {
                sb.append(" ");
            }
            if (remains-- > 0) {
                sb.append(" ");
            }
        }
        sb.append(list.get(m));
        res.add(sb.toString());
    }
}